### PR TITLE
Add polar, rot, and roa to can_do_properties_

### PR DESCRIPTION
**Pull request by bgpeyton**
## Description
~adds electric dipole polarizability, optical rotation, and electric dipole / magnetic dipole tensors to the `can_do_properties_` list so that they are correctly passed into `ret.extras` during a `run_qcshema` call. @loriab told you I'd ping you soon.~

both `omega` and `gauge` are picked up properly from Atomic Input keywords as well. 

UPDATE: `_can_do_properties` was really acting as more of a "default properties" list, so the logic in `run_qcschema` had to be changed instead. 

## Todos
- [x] `_can_do_properties` -> `_default_properties`
- [x] remove can-do filter from `run_qcschema` (kept default properties)
- [x] `run_qcschema` checks for `properties` list in `json['kwargs']` now instead of `json['model']`
- [x] new test  

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by PeterKraus** [comment]
The [qcschema test](https://github.com/psi4/psi4/blob/master/tests/pytests/test_psi4_qcschema.py) should probably be extended to check for this functionality.

**Comment by loriab** [comment]
If you haven't rebased recently, I wouldn't worry about the Windows error immediately, but Peter's right about tests. The test_psi4_qcschema.py is awkward to work with, so `tests/json/` is a good place to add tests. There's a lot of `run_json` in there to preserve backwards compatibility, but use `run_qcschema` instead.

**Comment by bgpeyton** [comment]
yes, I don't think there's anything testing things which end up in `extras` on the return, so I'll make a quick test. 

while I can accept the windows ctest fail, the failed `test_psi4_qcschema->test_psi4_scfproperty` for linux is a bit concerning. not really sure why it's trying to calculate SCF rotations when it's not in `function_kwargs`. it works fine when putting the properties list directly into the `model` of an `AtomicInput`, so it might be an issue in `qcng.util.model_wrapper` when translating the json data? I'll dig into it

**Comment by bgpeyton** [comment]
all kinds of weirdness going on, let's see if this checks all of the boxes. 

`test_psi4_scfproperty` was failing because the `properties` list was being overwritten, because `run_qcschema` was looking for properties in `model`. normally the calculation would have failed, but `_can_do_properties` was acting as a default list of properties (since it found none) and the test was only looking for a couple of things that happened to be in that list. 

to fix it, `run_qcschema` now correctly checks the kwargs for the property list. `_can_do_properties` was renamed to `_default_properties`, and the function no longer filters the properties through it (if the property is not available, psi4 should fail gracefully). once that was fixed, the "GRID_FIELD" and "GRID_ESP" properties needed to be removed from that test since they failed (could not find some `grid.dat`, I'll let someone else chime in on that one). 

also, added an additional test to hit the qc and wfn response vars (floats and arrays), both present in the `extras` return. I think this is the only test which checks the `extras` dictionary.

**Comment by bgpeyton** [comment]
pinging @loriab for review, not sure who else is intimately familiar with the schema harness (besides DGAS, not sure if he's reviewing lately?)

File: tests/json/schema-1-response/input.py
Comment (PeterKraus) [inline tests/json/schema-1-response/input.py:73]: Is this how `yapf` formats those?

File: psi4/driver/schema_wrapper.py
Comment (PeterKraus) [inline psi4/driver/schema_wrapper.py:549]: What happens if the asked-for property cannot be computed for whatever reason (wrong keyword as well as incompatible method and property)? Do we abort before any calculations are done, or only when we try to compute the properties?

File: tests/pytests/test_psi4_qcschema.py
Comment (PeterKraus) [inline tests/pytests/test_psi4_qcschema.py:118]: Uh, why don't we just save the json into a separate file and load that, instead of parsing this long string-as-json?

**Review by PeterKraus — CHANGES_REQUESTED** [review]
I am a little confused, as previously `can_do_properties_` also validated the requested properties, and now it trusts the asked-for properties are sensible.

File: psi4/driver/schema_wrapper.py
Comment (bgpeyton) [reply psi4/driver/schema_wrapper.py:549]: good question, it sort of depends on the method. for cc props,`proc.py` will print `"The following properties are not currently supported: ['PROPERTY']"`. any valid properties in the list are still computed. for scf, it prints `"SCF Property: Feature 'PROPERTY' is not recognized."` and suggests alternatives if possible, but no properties are computed, even the valid ones in the list. CI goes through the `proc_utils.oeprop_validator()`, algebraic diagrammatic construction does something similar to scf, and I don't think DFMP2 checks until somewhere later down the line.  

so as far as the Schema (and whatever is using it, like QCEng) is concerned, I guess the answer is "only when we try to compute the properties," then it's up to `proc.py` to have its own sanity checks depending on the module. I think it generally does a good job, so I think another layer of validating the method/prop combo in the schema wrapper isn't necessary.

File: tests/pytests/test_psi4_qcschema.py
Comment (bgpeyton) [reply tests/pytests/test_psi4_qcschema.py:118]: not gonna lie, I have no idea. I'm happy to do it if others think it's fine!

File: tests/json/schema-1-response/input.py
Comment (bgpeyton) [reply tests/json/schema-1-response/input.py]: ah, I forgot about `yapf`. formatted and pushed the change, thanks!

File: psi4/driver/schema_wrapper.py
Comment (PeterKraus) [reply psi4/driver/schema_wrapper.py:549]: OK, I am convinced that `schema_wrapper` may not be the best place to validate those inputs. 

I do think we need to make sure to "not crash" if it turns out a requested property (via qcschema or other spec) is not supported or is a typo. I want to avoid a situation where one requests a property that may not exist for a particular method, and then **nothing** gets returned instead of at least the stuff that has been computed already.

File: tests/pytests/test_psi4_qcschema.py
Comment (PeterKraus) [reply tests/pytests/test_psi4_qcschema.py:118]: I think the precedent of using external files for comparison in (py)tests is there - e.g. the `test_fchk_writer.py` looks into its own folder for fchk files. So I'd vote for doing it that way

Comment (hokru) [reply tests/pytests/test_psi4_qcschema.py:118]: There is a trick where you make a folder with the same name as the test and then use the `datadir` fixture at the top of `test_fchk_writer.py`.

**Comment by bgpeyton** [comment]
okay, tests finally passing. thanks to @PeterKraus for the help! if we agree that this isn't the place to ensure soft-fail for unsupported property kwargs, then I think this is ready!

**Comment by JonathonMisiewicz** [comment]
The big picture purpose of this PR is not clear to me. I understand that bullet point 4 is test coverage. Are the other three bullet points changing an interface, adding functionality, or pure code reorganization?

**Comment by bgpeyton** [comment]
@JonathonMisiewicz I would classify this as changing an interface. the wrapper will now push more properties into the `extra` slot of the return for a Schema run, rather than just throwing that information away (since the full wfn is not retrievable by schema). this adds flexibility to the interface w/ e.g. QCEngine and anything else using the wrapper, and also circumvents requiring the MolSSI to re-standardize the QCSchema for esoteric properties before we can return them.

File: psi4/driver/schema_wrapper.py
Comment (JonathonMisiewicz) [inline psi4/driver/schema_wrapper.py:545]: Do we still need a case change?

**Review by JonathonMisiewicz — COMMENTED** [review]
Okay, thanks. I have one comment on the actual code before I approve this.

As a general comment, Psi doesn't have the active reviewer base for the assumption "all the reviewers are familiar with this code" to hold. That's why it's always important for the description to supply enough context for the big picture purpose of the code to be clear, even to non-experts. That includes explicitly stating the problems/deficiencies in the current code that the PR solves.

File: psi4/driver/schema_wrapper.py
Comment (bgpeyton) [reply psi4/driver/schema_wrapper.py:545]: we shouldn't as far as I'm concerned -- the driver passes the kwargs along and the method runs as per usual, including input sanitization. (plus if there's a case issue, it would probably be better to catch it at the driver level so that it's fixed for any kind of input, not just JSON)

**Review by JonathonMisiewicz — APPROVED** [review]

**Review by PeterKraus — APPROVED** [review]
Thanks for getting the tests sorted.

File: tests/json/schema-1-response/input.py
Comment (loriab) [inline tests/json/schema-1-response/input.py:28]: is this test definitely running? it's using pytest fixtures and the `def test_*` collection pattern, but since it's in `tests/json/` rather than `tests/pytests/`, it's being executed by ctest and since everything's behind a `def`, I'm concerned it's a free pass. Maybe pytest is more powerful than I know, so just checking.

File: tests/json/schema-1-response/input.py
Comment (loriab) [inline tests/json/schema-1-response/input.py:77]: Is this a good candidate for converting several scalars to a single array? would need to extend some lists/tech in python_helpers.py

File: tests/json/schema-1-response/input.py
Comment (bgpeyton) [reply tests/json/schema-1-response/input.py:28]: good catch. CI doesn't run any of these json tests besides energy and gradient. but if you do force test these, the function inside `schema-1-response/input.py` does not get run. I can can pull it out of the function def, axe the fixtures, and put the reference data into the input directly instead of loading it from file. would that be fine?

File: tests/json/schema-1-response/input.py
Comment (bgpeyton) [reply tests/json/schema-1-response/input.py:77]: not familiar with the current tech, but probably. extension of the helper functions might be better suited for a separate PR, though.

File: tests/json/schema-1-response/input.py
Comment (loriab) [reply tests/json/schema-1-response/input.py:28]: that would be fine. or you can move the file and ref file to tests/pytests/. either way

Comment (bgpeyton) [reply tests/json/schema-1-response/input.py]: done. to make sure it was _actually_ testing, I tried putting a "bad" value in and ran the test, and it failed as expected.

File: tests/json/schema-1-response/input.py
Comment (loriab) [inline tests/json/schema-1-response/input.py:321]:
```suggestion
json_ret = psi4.schema_wrapper.run_qcschema(json_data).dict()
```

File: tests/json/schema-1-response/input.py
Comment (loriab) [inline tests/json/schema-1-response/input.py:319]:
```suggestion
expected_response = {k: (np.asarray(v) if isinstance(v, List) else v) for k, v in expected_response.items()}
```

File: tests/pytests/test_psi4_qcschema.py
Comment (loriab) [inline tests/pytests/test_psi4_qcschema.py:164]: thanks, it's certainly tidier this way.

File: psi4/driver/schema_wrapper.py
Comment (loriab) [inline psi4/driver/schema_wrapper.py:545]: I guess we won't catch properties like GRID_ESP that can't be returned, but as this is mostly for CC, perhaps that's ok.

File: tests/json/schema-1-response/input.py
Comment (bgpeyton) [reply tests/json/schema-1-response/input.py:319]: wow, that's really nifty

File: psi4/driver/schema_wrapper.py
Comment (bgpeyton) [reply psi4/driver/schema_wrapper.py:545]: yes, there are some edge cases that will probably slip through. maybe we could collect them and patch them through in another way, if anyone is interested in the future.

**Review by loriab — APPROVED** [review]
thanks! sorry it's taken me so long to grasp the problem/sol'n in this PR.

File: tests/json/schema-1-response/input.py
Comment (bgpeyton) [reply tests/json/schema-1-response/input.py]: `List` needs to be `list`, oops. now test is passing again 👍

**Pull request merged**
